safekeeper/src/lib.rs

```diff
@@ -91,11 +91,62 @@ impl SafeKeeperConf {
                 .parse()
                 .expect("failed to parse default broker endpoint"),
             broker_keepalive_interval: Duration::from_secs(5),
-            backup_runtime_threads: None,
             wal_backup_enabled: true,
             auth: None,
             heartbeat_timeout: Duration::new(5, 0),
             max_offloader_lag_bytes: defaults::DEFAULT_MAX_OFFLOADER_LAG_BYTES,
         }
     }
 }
+
+// Tokio runtimes.
```

Comment:
Why add all these runtimes?

One would be enough. This combination also has maximum thread count of 6 * 512 assuming they use `spawn_blocking` directly or indirectly. A single runtime should be very much enough if all file access apart from initial loading is done with tokio's functionality.

A single runtime would allow testing with a single thread runtime, which is an easy way to spot any deadlocks in a system. For this huge amount of worker threads it will be very difficult.

Reply:
Looks like the `async fn main` is already driven by a `current_thread` runtime, which is great. If all of these were to be removed, and only `tokio::spawn` used, you'd get the deadlock "testing" right away.

Reply:
https://github.com/neondatabase/neon/pull/4124/commits/ed4fd15b7cb122f5232787b8a8750e594e84a699 has two test failures, unsure yet if they are transient.

Reply:
Those failures are "expected":

> Since metrics now can be delayed up to a scrape period due to clutch described above, tests relying on them fail.

Not sure yet what is the good fix for the metrics.